Adding latitudinal variation to Tidal heating model

hyunseong96 opened the pull request:
I open a PR to add latitudinal variation of tidal heating. There are three new parameters: ‘Use latitudinal variation of strain rate’, ‘Maximum tidal strain rate’ and ‘Minimum tidal strain rate’.

The bool parameter ‘Use latitudinal variation of strain rate’ will be used to decide whether to use latitudinal variation of tidal heating. I made AssertThrow in initialize() because model should be spherical geometry to set true for ‘Use latitudinal variation of strain rate’, as Box does not have latitudinal variation. However, as far as I know, ASPECT does not have latitude variation on 2D, so the variation is only found in 3D. If it is better to add additional AssertThrow for 2D, I will add it.

The latitudinal variation of tidal strain rate is simplified with cosine function and is between ‘Maximum tidal strain rate’ and ‘Minimum tidal strain rate’. You can check the solid line in the figure below (Nimmo et al., 2007):
<img width="400" alt="image" src="https://github.com/user-attachments/assets/729c358b-0958-4cdc-aab1-53e45f80cfb7" />

To calculate the variation, I added variable 'local_tidal_strain_rate' to calculate strain rate at certain polar angle (theta). If ‘Use latitudinal variation of strain rate’ is true, polar angle (theta) is calculated, then local_strain_rate is found. If false, ‘Constant tidal strain rate’ is used, just as tidal heating affected by viscosity.

The test file is the case where tidal heating only exerts without convection and conduction. Geometry is 3D spherical shell with 360-degree opening angle. Viscosity is constant as 1e14 $Pa{\cdot}s$ in simpler material model. This is the temperature from 100K everywhere after 1Myr of heating. Analytical values are 2.85946709e+02 K and 1.65676318e+02 K at the pole and equator, respectively. These values do not match perfectly within decimal places that ASPECT writes but align in some order with calculated values through the code, 2.85952323e+02 K and 1.65675477e+02 K at poles and equators, respectively. I personally think this error, within ~<2e-3, comes from spherical geometry and mesh resolution.

I tried with box model with true for 'Use latitudinal variation of strain rate' and checked that AssertThrow works to  send ExcMessage. Also, tidal heating works well when 'Use latitudinal variation of strain rate' is false.

<img width="300" alt="image" src="https://github.com/user-attachments/assets/79e4b4fb-d9f2-4c6c-9d83-a5a7e14e5cc6" />
<img width="300" alt="image" src="https://github.com/user-attachments/assets/20263924-dc1d-41b0-89bf-a1f0ad3c0bd3" />

gassmoeller reviewed (commented):
Looks already pretty good, but I have some suggestions for improvements. Address them and let me know when you have questions or are ready for another review.
Also add a changelog entry (or modify your previous one).

hyunseong96 commented on tests/tidal_heating_latitude.prm:
```diff
@@ -0,0 +1,137 @@
+  #  set Radius repetitions = 1
+  #  set Latitude repetitions = 15
+  #  set Longitude repetitions = 15
+  #end
```
@gassmoeller I tried 90-degree opening angle to 3D spherical shell. but I kept getting this error. It worked fine with 360 degree.

```
../aspect tidal_heating_latitude.prm
-----------------------------------------------------------------------------
--                             This is ASPECT                              --
-- The Advanced Solver for Planetary Evolution, Convection, and Tectonics. --
-----------------------------------------------------------------------------
--     . version 3.1.0-pre (th_latitude2, 2c030ecad)
--     . using deal.II 9.6.1
--     .       with 32 bit indices
--     .       with vectorization level 1 (SSE2, 2 doubles, 128 bits)
--     . using Trilinos 14.4.0
--     . using p4est 2.3.6
--     . using Geodynamic World Builder 1.0.0
--     . running in DEBUG mode
--     . running with 1 MPI process
-----------------------------------------------------------------------------

-----------------------------------------------------------------------------
-- For information on how to cite ASPECT, see:
--   https://aspect.geodynamics.org/citing.html?ver=3.1.0-pre&mf=1&sha=2c030ecad&src=code
-----------------------------------------------------------------------------
Number of active cells: 24 (on 2 levels)
Number of degrees of freedom: 1,277 (915+57+305)


--------------------------------------------------------
An error occurred in line <829> of file </Users/hyunseong/aspect/source/simulator/helper_functions.cc> in function
    double aspect::Simulator<3>::normalize_pressure(LinearAlgebra::BlockVector &) const [dim = 3]
The violated condition was: 
    area > 0
Additional information: 
    While computing the average pressure, the area/volume to integrate
    over was found to be zero or negative. This indicates that no
    appropriate surface faces were found, which is typically the case if
    the geometry model is not set up correctly.

Stacktrace:
-----------
#0  2   aspect-debug                        0x000000010132ff74 _ZNK6aspect9SimulatorILi3EE18normalize_pressureERN6dealii16TrilinosWrappers3MPI11BlockVectorE + 3468: 2   aspect-debug                        0x000000010132ff74 _ZNK6aspect9SimulatorILi3EE18normalize_pressureERN6dealii16TrilinosWrappers3MPI11BlockVectorE 
#1  3   aspect-debug                        0x00000001013432b0 _ZN6aspect9SimulatorILi3EE30compute_initial_pressure_fieldEv + 1316: 3   aspect-debug                        0x00000001013432b0 _ZN6aspect9SimulatorILi3EE30compute_initial_pressure_fieldEv 
#2  4   aspect-debug                        0x0000000101a0ab8c _ZN6aspect9SimulatorILi3EE3runEv + 796: 4   aspect-debug                        0x0000000101a0ab8c _ZN6aspect9SimulatorILi3EE3runEv 
#3  5   aspect-debug                        0x00000001019787e0 _Z13run_simulatorILi3EEvRKNSt3__112basic_stringIcNS0_11char_traitsIcEENS0_9allocatorIcEEEES8_bbbb + 1076: 5   aspect-debug                        0x00000001019787e0 _Z13run_simulatorILi3EEvRKNSt3__112basic_stringIcNS0_11char_traitsIcEENS0_9allocatorIcEEEES8_bbbb 
#4  6   aspect-debug                        0x000000010197772c main + 2240: 6   aspect-debug                        0x000000010197772c main 
#5  7   dyld                                0x00000001988160e0 start + 2360: 7   dyld                                0x00000001988160e0 start 
--------------------------------------------------------

[hyunseong-MacBook-Air:05784] *** Process received signal ***
[hyunseong-MacBook-Air:05784] Signal: Abort trap: 6 (6)
[hyunseong-MacBook-Air:05784] Signal code:  (0)
[hyunseong-MacBook-Air:05784] [ 0] 0   libsystem_platform.dylib            0x0000000198bcf584 _sigtramp + 56
[hyunseong-MacBook-Air:05784] [ 1] 0   libsystem_pthread.dylib             0x0000000198b9ec20 pthread_kill + 288
[hyunseong-MacBook-Air:05784] [ 2] 0   libsystem_c.dylib                   0x0000000198aaba20 abort + 180
[hyunseong-MacBook-Air:05784] [ 3] 0   libdeal_II.g.9.6.1.dylib            0x00000001180013b4 _ZN6dealii18deal_II_exceptions9internals5abortERKNS_13ExceptionBaseE + 164
[hyunseong-MacBook-Air:05784] [ 4] 0   aspect-debug                        0x0000000100e60ed8 _ZN6dealii18deal_II_exceptions9internals20issue_error_noreturnINS_18StandardExceptions10ExcMessageEEEvNS1_17ExceptionHandlingEPKciS7_S7_S7_T_ + 224
[hyunseong-MacBook-Air:05784] [ 5] 0   aspect-debug                        0x000000010132ff74 _ZNK6aspect9SimulatorILi3EE18normalize_pressureERN6dealii16TrilinosWrappers3MPI11BlockVectorE + 3468
[hyunseong-MacBook-Air:05784] [ 6] 0   aspect-debug                        0x00000001013432b0 _ZN6aspect9SimulatorILi3EE30compute_initial_pressure_fieldEv + 1316
[hyunseong-MacBook-Air:05784] [ 7] 0   aspect-debug                        0x0000000101a0ab8c _ZN6aspect9SimulatorILi3EE3runEv + 796
[hyunseong-MacBook-Air:05784] [ 8] 0   aspect-debug                        0x00000001019787e0 _Z13run_simulatorILi3EEvRKNSt3__112basic_stringIcNS0_11char_traitsIcEENS0_9allocatorIcEEEES8_bbbb + 1076
[hyunseong-MacBook-Air:05784] [ 9] 0   aspect-debug                        0x000000010197772c main + 2240
[hyunseong-MacBook-Air:05784] [10] 0   dyld                                0x00000001988160e0 start + 2360
[hyunseong-MacBook-Air:05784] *** End of error message ***
zsh: abort      ../aspect tidal_heating_latitude.prm

```

gassmoeller commented on include/aspect/heating_model/tidal_heating.h:
```diff
@@ -74,10 +83,25 @@ namespace aspect
          * tidal frequency = tidal_frequency
          * elastic shear modulus = elastic_shear_modulus
+         *
+         * latitudinal_variation is bool that used to decide whether latitudinal variation would be considered.
+         * If true, strain rate varies by latitude having maximum at the poles and minimum at the equator.
```
this comment is now out of date

gassmoeller commented on include/aspect/heating_model/tidal_heating.h:
```diff
@@ -74,10 +83,25 @@ namespace aspect
         double tidal_frequency;
         double elastic_shear_modulus;
         double constant_tidal_strain_rate;
+        /**
+         * Specify the distribution of time-averaged tidal strain rate.
+         */
+        enum CustomTidalHeating
+        {
+          constant,
+          latitudinal_variation
+        } custom_strain_rate;
```
just for readability, please put empty lines before the comment in line 95 and after the variable in line 102 (i.e. around the customtidalheating enum)

also maybe rename the enum to `StrainRateDistribution` and the variable to `strain_rate_distribution`

gassmoeller commented on source/heating_model/tidal_heating.cc:
```diff
@@ -21,30 +21,72 @@
+    void
+    TidalHeating<dim>::initialize ()
+    {
+      if (latitudinal_variation)
```
this should now be:
```suggestion
      if (strain_rate_distribution == latitudinal_variation)
```

gassmoeller commented on source/heating_model/tidal_heating.cc:
```diff
@@ -85,6 +127,23 @@ namespace aspect
+                             "If 'latitudinal variation', 'Maximum tidal strain rate' and 'Minimum tidal strain rate' are used.");
+          prm.declare_entry ("Maximum tidal strain rate", "2.81e-10",
+                             Patterns::Double (0),
+                             "Maximum time-averaged tidal strain rate by lunar diurnal tide that is at the poles. "
```
```suggestion
                             "Maximum time-averaged tidal strain rate by lunar diurnal tide at the poles. "
```

gassmoeller commented on source/heating_model/tidal_heating.cc:
```diff
@@ -85,6 +127,23 @@ namespace aspect
+                             "Units: 1/s.");
+          prm.declare_entry ("Minimum tidal strain rate", "1.67e-10",
+                             Patterns::Double (0),
+                             "Minimum time-averaged tidal strain rate by lunar diurnal tide that is at the equator. "
```
```suggestion
                             "Minimum time-averaged tidal strain rate by lunar diurnal tide at the equator. "
```

gassmoeller commented on source/heating_model/tidal_heating.cc:
```diff
@@ -85,6 +127,23 @@ namespace aspect
+          prm.declare_entry ("Maximum tidal strain rate", "2.81e-10",
+                             Patterns::Double (0),
+                             "Maximum time-averaged tidal strain rate by lunar diurnal tide that is at the poles. "
+                             "This parameter will be used when 'Use latitudinal variation of strain rate' is true. "
```
this comment line needs an update

gassmoeller commented on source/heating_model/tidal_heating.cc:
```diff
@@ -85,6 +127,23 @@ namespace aspect
+          prm.declare_entry ("Minimum tidal strain rate", "1.67e-10",
+                             Patterns::Double (0),
+                             "Minimum time-averaged tidal strain rate by lunar diurnal tide that is at the equator. "
+                             "This parameter will be used when 'Use latitudinal variation of strain rate' is true. "
```
this comment line needs an update

gassmoeller reviewed (commented):
Yes, very close. Just a few comments left.

bangerth commented:
Certainly agreement with analytical values to 4 digits of accuracy can be considered "exact" :-)

bangerth commented on source/heating_model/tidal_heating.cc:
```diff
@@ -21,30 +21,72 @@
+      if (latitudinal_variation)
+        {
+          // Only spherical geometries will be accepted to have heating rate varying by latitude.
+          AssertThrow(Plugins::plugin_type_matches<GeometryModel::Chunk<dim>>(this->get_geometry_model()) ||
+                      Plugins::plugin_type_matches<GeometryModel::TwoMergedChunks<dim>>(this->get_geometry_model()) ||
+                      Plugins::plugin_type_matches<GeometryModel::EllipsoidalChunk<dim>>(this->get_geometry_model()) ||
+                      Plugins::plugin_type_matches<GeometryModel::Sphere<dim>>(this->get_geometry_model()) ||
+                      Plugins::plugin_type_matches<GeometryModel::SphericalShell<dim>>(this->get_geometry_model()),
+                      ExcMessage("Latitudinal variation of strain rate can only be used with 3-dimensional geometry models that "
+                                 "have either a spherical or ellipsoidal natural coordinate system."));
```
From your description, I think you also want to check that `dim==3`.

hyunseong96 replied:
@bangerth I added this! :)

hyunseong96 commented:
@gassmoeller I have updated comments! 
Also in 2D, line `const double theta = Utilities::Coordinates::cartesian_to_spherical_coordinates(position)[dim-1];` makes variable `theta` an azimuthal angle.
So I added `dim==3` to AssertThrow in initialize(), as @bangerth mentioned, not to have variation in 2D.
If latitudinal variation is needed in 2D, I would add through other PR.

bangerth commented:
Can you combine the commits into one?

bangerth commented:
I really want to merge this, but there's currently an error in one of the testers:
```
/home/runner/work/aspect/aspect/source/heating_model/tidal_heating.cc:47:116: error: suggest parentheses around ‘&&’ within ‘||’ [-Werror=parentheses]
   47 |                       Plugins::plugin_type_matches<GeometryModel::SphericalShell<dim>>(this->get_geometry_model()) &&
      |                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
   48 |                       dim==3,
      |                       ~~~~~~                                                                                        
/home/runner/work/aspect/aspect/source/heating_model/tidal_heating.cc: In instantiation of ‘void aspect::HeatingModel::TidalHeating<dim>::initialize() [with int dim = 3]’:
/home/runner/work/aspect/aspect/source/heating_model/tidal_heating.cc:189:5:   required from here
/home/runner/work/aspect/aspect/source/heating_model/tidal_heating.cc:47:116: error: suggest parentheses around ‘&&’ within ‘||’ [-Werror=parentheses]
```
You may want to address that.

bangerth commented:
You forgot to run the indentation script, and so the indentation is now wrong after you added the parentheses :-( Once you fix this, would you mind squashing everything into one commit again?
